### PR TITLE
Add ties.method's first, last, random, and dense to row|colRanks

### PR DESCRIPTION
**Pull request by bkmontgom**
The goal is to be compatible with base::rank
Add ties.method's "first", "last", "random"
Change the order of the ties.method options to match base::rank
So new default ties.method is "average"
Also add "dense" option as in data.table::frank

### CODE REVIEW

**Comment by HenrikBengtsson** [comment]
This is awesome, thank you for doing this.  I'll comment more in Issue #142, but we need to transition slowly, so could you please update so the default for `ties.method` is unchanged.  One way is to just prepend the current default:
```r
                   ties.method = c("max", "average", "first", "last", "random",
                                     "max", "min", "dense")
```
It'll somewhat signal there's something "going on" with that argument.

When that is in place, I'll try to run ASAN/UBSAN tests too.  (I haven't got that set up on my current computer, so it'll take some time. Tried do to it via r-hub but that eventually timed out.  I might look into using https://hub.docker.com/r/rocker/r-devel-ubsan-clang via Docker/Singularity.)

**Comment by bkmontgom** [comment]
I'll put the default back to "max".  I like the idea of just prepending it to the front of the argument list.  I'll leave the "If missing(ties.method) == TRUE.." part to you.

**Comment by HenrikBengtsson** [comment]
Thanks for this - I still haven't had time to setup the ASan/UBsan test system.

I see you start fixing other functions too - could you please commit/push them to a separate branch of yours and create a separate PR for those?  That'll help me to focus on a smaller set of the code - the larger the code base is to review is, the harder it will be to test/accept.  Thxs

**Comment by bkmontgom** [comment]
Okay.  Sorry I'm a git novice.

**Comment by HenrikBengtsson** [comment]
FYI, I'm going to start to work my way through this PR. I'll add inline code-review comments as I find potentials issues. Please hold back until I give a "go ahead" here before responding to those code-review comments; I might revise some of them later.

File: tests/rowRanks.R [hunk 1]
```diff
@@ -3,57 +3,130 @@ library("matrixStats")
...
+  for (ties in c("random")) {
+    cat(sprintf("ties.method = %s\n", ties))
+    # rowRanks():
+    y1 <- matrixStats::rowRanks(x[[kk]], ties.method = ties)
+    y2min <- t(apply(x[[kk]], MARGIN = 1L, FUN = rank, na.last = "keep",
+                     ties.method = "min"))
+    y2max <- t(apply(x[[kk]], MARGIN = 1L, FUN = rank, na.last = "keep",
+                     ties.method = "max"))
+    stopifnot(all(y1 >= y2min, na.rm = TRUE) && all(y1 <= y2max, na.rm = TRUE))
+    y1list <- Map(function(...) matrixStats::rowRanks(x[[kk]], ties.method = ties), 1:10000)
```
[the remark this answers is not in the document]
Comment (HenrikBengtsson) [reply tests/rowRanks.R]: Using 10,000 samples seems to work also when using Valgrind; the total build time on Travis CI is ~10-15 mins which is way below the maximum 50 minutes.

File: src/rowRanksWithTies_lowlevel_template.h [hunk 2]
```diff
@@ -150,15 +155,37 @@ void METHOD_NAME_ROWS_COLS(ARGUMENTS_LIST) {
...
     // Calculate the ranks.
+    firstTie = 0;
+    aboveTie = 1;
+    dense_rank_adj = 0;
     for (jj=0; jj <= lastFinite;) {
-      firstTie = jj;
+      if (TIESMETHOD == 'd') {
+        dense_rank_adj += (aboveTie - firstTie - 1);
+      }
+      firstTie = jj - dense_rank_adj;
```
[the remark this answers is not in the document]
Comment (HenrikBengtsson) [reply src/rowRanksWithTies_lowlevel_template.h]: Done in commit fadd55b19

**Comment by HenrikBengtsson** [comment]
Everything is now merged into the develop branch.  Thanks a bunch for this contribution.

**Comment by bkmontgom** [comment]
Great!  You're welcome.  All good comments.  I really struggled with how to test "random".

Sent from Yahoo Mail on Android 
 
  On Sun, Jul 21, 2019 at 8:03 AM, Henrik Bengtsson<notifications@github.com> wrote:   
Everything is now merged into the develop branch. Thanks a bunch for this contribution.

**Comment by HenrikBengtsson** [comment]
matrixStats 0.55.0 is now on CRAN. Thanks for your contribution.

**Pull request merged**